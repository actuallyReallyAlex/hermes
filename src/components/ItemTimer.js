import React, { useState, useEffect } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import {
  setTimerRunning,
  clearItems,
  refreshItems
} from '../redux/actions/world'
import { itemTimerLogic } from '../util'
import { Box, Heading } from 'grommet'

const ItemTimer = ({ handleTimerStarted, handleTimerStopped, world }) => {
  const [timeLeft, setTimeLeft] = useState(null)

  useEffect(
    () =>
      itemTimerLogic(
        world,
        setTimeLeft,
        handleTimerStarted,
        handleTimerStopped
      ),
    // eslint-disable-next-line
    [world.isTimerRunning]
  )

  return (
    <Box>
      <Heading level="3">Item Timer</Heading>
      <span>{timeLeft}</span>
    </Box>
  )
}

ItemTimer.propTypes = {
  handleTimerStopped: PropTypes.func.isRequired,
  handleTimerStarted: PropTypes.func.isRequired
}

const mapStateToProps = ({ world }) => ({ world })

const mapDispatchToProps = dispatch => ({
  handleTimerStarted: () => dispatch(setTimerRunning(true)),
  handleTimerStopped: () => {
    // * Clear all items from planets
    dispatch(clearItems())
    // * Put new items on planets
    dispatch(refreshItems())
    // * Tell Redux the timer is no longer running
    dispatch(setTimerRunning(false))
  }
})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ItemTimer)
