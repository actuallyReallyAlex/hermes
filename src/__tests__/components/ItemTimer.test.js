/**
 * @jest-environment node
 */

import React from 'react'
import ItemTimer from '../../components/ItemTimer'
import { mount, shallow } from 'enzyme'
import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store'

const middlewares = []
const mockStore = configureStore(middlewares)

const defaultState = {
  ship: {
    cargo: [],
    location: {
      name: null,
      value: null
    }
  },
  ui: {
    view: 'Ship'
  },
  user: {
    cash: 100
  },
  world: {
    isTimerRunning: false,
    planets: []
  }
}

const rerenderShallow = customState =>
  shallow(
    <Provider store={mockStore(customState ? customState : defaultState)}>
      <ItemTimer />
    </Provider>
  )

const rerenderMount = customState =>
  mount(
    <Provider store={mockStore(customState ? customState : defaultState)}>
      <ItemTimer />
    </Provider>
  )

describe('<ItemTimer />', () => {
  it('Should render the <ItemTimer /> component.', () => {
    const wrapper = rerenderShallow()
    expect(wrapper.html()).toMatchSnapshot()
  })
})
