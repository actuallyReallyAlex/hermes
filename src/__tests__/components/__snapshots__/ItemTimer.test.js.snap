// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ItemTimer /> Should render the <ItemTimer /> component. 1`] = `
<DocumentFragment>
  <div
    class="StyledBox-sc-13pk1d4-0 fWrrBh"
  >
    <h3
      class="StyledHeading-sc-1rdh4aw-0 ftOecI"
    >
      Item Timer
    </h3>
    <span />
  </div>
</DocumentFragment>
`;
